sherpa/astro/ui/utils.py

```diff
@@ -8727,17 +8740,26 @@ def fake_pha(self, id, arf, rmf, exposure, backscal=None, areascal=None,
         ...          grouping=grp, quality=qual, grouped=True)
         >>> save_pha('sim', 'sim.pi')
 
+        Sometimes, the background dataset is noisy because there are not
+        enough photons in the background region. In this case, the background
+        model can be used to generate the photons that the background
+        contributes to the source spectrum. To do this, a background model
+        must be passed in. This model is then convolved with the ARF and RMF
+        (which must be set before) of the default background data set:
+
+        >>> set_bkg_source('sim', 'const1d.con1')
+        >>> load_arf('sim', 'bkg.arf.fits', bkg_id=1)
+        >>> load_rmf('sim', 'bkg_rmf.fits', bkg_id=1)
+        >>> fake_pha('sim', arf, rmf, texp, backscal=bscal, bkg='model',
+        ...          grouping=grp, quality=qual, grouped=True)
+        >>> save_pha('sim', 'sim.pi')
         """
         id = self._fix_id(id)
-        d = sherpa.astro.data.DataPHA('', None, None)
+
         if id in self._data:
             d = self._get_pha_data(id)
         else:
-            # Make empty header OGIP compliant
-            # And add appropriate values to header from input values
-            d.header = dict(HDUCLASS="OGIP", HDUCLAS1="SPECTRUM",
-                            HDUCLAS2="TOTAL", HDUCLAS3="TYPE:I",
-                            HDUCLAS4="COUNT", HDUVERS="1.1.0")
+            d = sherpa.astro.data.DataPHA('', None, None)
```

Comment:
My current thinking is that we shouldn't allow a user to create an empty `Data` object - we should make them supply this (and ideally make the independent/depentent data immutable [I think this is actually a step too far] or at least don't change their size). However, that's not for this PR...

What would make the above easier to do (at least for the `DataPHA` case is if we can re-arrange the logic of this method do that we move the `DataPHA` call until after we've created `rmf0`, so we can do something like (I haven't checked this is sufficient or if it isn't actually this easy because of the information flow).

```python
if id in self._data:
    d = self._get_pha_data(id)
    if len(d.channel) != rmf0.detchans:
        raise Oops

    for resp_id in d.resonse_ids:
        d.delete_response(resp_id)

    ... maybe something more

else:
    chans = sao_arange(1, rmf0.detchans)
    counts = chans * 0
    d = DataPHA('faked', chans, counts) 
```

We might also want to stop digging around in the internals of the Session class and we should probably just use try/except (although I don't like having so much code unrelated to the exception handler running within it):

````
try:
    d = self._get_pha_data(id)
    ... check number of channels
    for d.response_ids:
        ...
except IdentifierErr:
    .... need to create the object
```

Reply:
I'm not sure I share the concern about "empty" Data objects. The object is not returned to the caller until it is complete. So, the user never sees the empty object. I guess we can also collect all the pieces (counts, channels, exposure time) in independent variable and then create the DataPHA once it's all ready. The user won't see the difference. I'll look if that's easy to do.

Reply:
Well, the empty DataPHA object isn't a problem for the user of this code. It is a problem if I go and change DataPHA.__init__ to ensure that channel and counts are set - this is, I believe, the only place in our code where that would fail, so I'm eager to tweak the code here!

Reply:
We need to discuss and settle this before I can do any further work:
For the OO interface, I've settled on an interface where user sets up the DataPHA object with arf, rmf, exposure time etc - essentially all but the channels and counts. Channels and counts will then be filled in by `fake_pha`. That only works if the user can make an empty `DataPHA` object. The alternative is what the old UI method did - you pass in arf, rmfs, exposure time, etc. as separate parameters and then the methods puts them all into a the `DataPHA` object. I think that's cumbersome and complex - just look at the old `fake_pha` UI code because that code needs to duplicate  the load_arf/set_arf code to check if the arf is an object or a filename. If it's a filename it check if that file exists, ... I don't think "checking is a file exists" should be in the purview of `fake_pha`. It gets even more complex if we want to set up background, multiple backgrounds, or multiple arfs/rmfs (the goal of this proposal). Now, I also need to replicate the interface (and some of the code) for `set_multi_arf`, `set_multi_rmf`, and `load_background` . Why would I want to do that? We do have methods that do that already. Then, we could make and object that holds the ARF, RMF, background, exposure time, OGIP compliant header etc. so that we can pass it around, e.g. from the UI layer to the `fake_pha` in the OO layer - but we do have such an object already: `DataPHA`.

So, I think that `fake_pha` should be limited to actually do the faking and leave the setup (loading, setting arf, rmf, background, exposure time) to other existing routines, at least in the OO layer. In the UI layer, we keep the legacy interface, but, since that calls the OO layer to do the actual faking, we do need to set up the empty `DataPHA` object.

I also think that conceptually, that is the right way to think about it: We are creating a new dataset, so we should be able to create a new `DataPHA` object. The only thing that's unusual (=different from reading a fits file) is that we have the meta information first and the channels and counts later. However, that's how simulations work: We set up the conditions and then we run the simulation to get the counts.

In summary: I think any way that avoids making an "empty" `DataPHA` is going to be very cumbersome in some way or another. What we can do is to load the arf first and then create the `DataPHA` with some dummy entries to avoid the `None,None` in the signature. However, that's mostly cosmetic and does not address the deeper point that we'll only know the counts **after** we've set up everything else.

Reply:
I haven't read this thoughtfully but I think we can have the best of both worlds - we can allow users / the code to change the counts array as long as it is of the correct size. So, we could for the DataPHA case require the channel array to be sent to `__init__` and it can set up an empty counts array if not set. This wouldn't need to be done before this PR is done.

What I really would like is for our code to be able to ensure that things have the correct sizes which means we then we can be more confident that downstream code isn't going to fall over because we've just tried to combine the counts and syserror columns and they don't match. 

However, saying all this is a bit academic as I need to look at #1197 as I suspect my idealism is about to be dashed by reality.

Reply:
Let's think about the OO interface for the moment, because that's cleaner (the UI has code around it to set up things with the right dataset numbers etc.), so in OO, I'd like to do:
```
fakes = DataPHA(None, None)
fakes.exposure = 5000
fakes.load_arf('XXX.arf')
fakes.load_rmf('XXX.rmf')
fake_pha(fakes)
```

If, instead, I need to make fake channel and count arrays then it goes something like this:
```
arf = load_arf('XXX.arf')
counts = np.zeros_like(arf.channels)
fakes = DataPHA(arf.channels, counts)
fakes.exposure = 5000
fakes.load_arf('XXX.arf')
fakes.load_rmf('XXX.rmf')
fake_pha(fakes)
```
(conceptually, I don't actually know on top of my head if `.counts` is the right thing). Why am I loading the arf twice? What's the real purpose of that intermediate "counts" array? 
I'm not saying it can't be done this way, I'm just saying that it's not "free" as in "it complicates the API", so I want to make absolutely sure that that's worth the trouble.

Reply:
> The TL;DR of this comment is that my hopes are likely not going to work (that is, forcing validation checks when we make changes to the DataPHA object rather than when we evaluate the object to get data out) and so we should just ignore these thoughts and carry on with this PR.

To be picky you can't use the ARF but the RMF (arf's do not have a channels attribute and do not contain the correct number of rows in geeral), so it would be something like

```python
rmf = unpack_rmf('XXX.rmf')
counts = np.arange(rmf.offset, rmf.detchans + rmf.offset)
fakes = DataPHA(counts, np.zeros_like(counts))
fakes.exposure = 5000
fakes.set_rmf(rmf)
fakes.set_arf(unpack_arf('XXX.arf'))
fake_pha(fakes)
```

(here using existing features not some of the work you've been playing with).

An important part of the PHA concept is the range of channels - which can be larger than the actual channels argument, but this is not something we really support at the moment. What I really want is to be able to validate arguments - such as the quality array - have the correct length when set. This is easier if we force the channels to be set when `DataPHA` is called, but as you show it's a bit annoying. What we could do is allow the values to be None but as soon as something is set that defines the channels (which I was thinking would be setting the channels but in your example it's setting the RMF) then we can enforce it. I now have serious concerns about my whole idea since we have not really focused on ARF/RMF ordering but if we were to do this we'd need to enforce RMF first (unless it's a square RMF matrix so we could set ARF first).

Reply:
Thinking about this more: I totally understand what @DougBurke is coming from: Our lives as coders get a lot simpler if we can validate an object in the `__init__` and after that we can assume that everything fits. However, for the simulation, that's a wrong way around - and simulations are an important use case for us.
So, I think we are stuck with checking consistency every time the counts, channels, arf, rmfs are changed e.g. in the the set_arf, set_rmf, or in a property setter for counts and channels.

Bottom line: Let's, as @DougBurke suggests, "ignore these thoughts [at least for now] and carry on with the PR"